python/sglang/backend/openai.py

```diff
@@ -229,6 +235,112 @@ def spec_pattern_match(self, comp):
+
+        def function_to_json_schema(func):
+            signature = inspect.signature(func)
+            parameters = signature.parameters
+            func_schema = {
+                "type": "function",
+                "function": {
+                    "name": func.__name__,
+                    "parameters": {
+                        "type": "object",
+                        "properties": {
+                            param.name: {
+                                "type": convert_param_type(
+                                    str(param.annotation)
+                                    .replace("<class '", "")
+                                    .replace("'>", "")
+                                )
+                            }
+                            for param in parameters.values()
+                        },
+                    },
+                },
+            }
+            return func_schema
+
+        tools_to_use = []
+        if tools:
+            tools_to_use = [
+                function_to_json_schema(tool_to_use) for tool_to_use in tools
+            ]
+        cur_tool_choice = "auto"
+        if tool_choice:
+            cur_tool_choice = (
+                tool_choice
+                if tool_choice in ["auto", "required", "none"]
+                else {"type": "function", "function": {"name": tool_choice}}
+            )
+
+        # TODO: "Never mention what tools you use." or provide a system prompt input argument
+        response = self.client.chat.completions.create(
+            model=self.model_name,
+            messages=s.messages_,
+            tools=tools_to_use,
+            tool_choice=cur_tool_choice,
+            **self.spec_kwargs,
+        )
+        response_message = response.choices[0].message
+        tool_calls = response_message.tool_calls
+        # Check if the model wanted to call a function
+        ret_messages = []
+        if tool_calls:
+            # Call the function
+            # Note: the JSON response may not always be valid; be sure to handle errors
+            available_functions = {}
+            for tool in tools:
+                available_functions[tool.__name__] = tool
+            ret_messages.append(response_message)
+            # Send the info for each function call and function response to the model
+            for tool_call in tool_calls:
+                function_name = tool_call.function.name
+                function_to_call = available_functions[function_name]
+                function_args = json.loads(tool_call.function.arguments)
+                function_response = function_to_call(**function_args)
+                ret_messages.append(
+                    {
+                        "tool_call_id": tool_call.id,
+                        "role": "tool",
+                        "name": function_name,
+                        "content": str(function_response),
+                    }
+                )
+        return ret_messages
```

Comment:
I think it is better to put the logic of real function call into the interpreter, so that it can be reused when we develop the feature for local models.
And remember to handle the logic of appending `s.messages_` and `s.text_` in the interpreter.

Reply:
Makes sense, returning just the function call messages here so we can do real function call separately.